service/storage_service.cc

```diff
@@ -278,6 +278,8 @@ void storage_service::prepare_to_join(std::vector<inet_address> loaded_endpoints
     if (get_replace_tokens().size() > 0 || get_replace_node()) {
```

Comment:
General question about the commit (and commit message): it says that if we're replacing a node with different IP, we should update tokens AFTER replace is done, in contrast to the other case (replacing with the same IP), where we update tokens BEFORE.

But I don't understand WHY. The commit message and preferably comments inside the code should explain it.

Reply:
Say the node to be replaced is 127.0.0.3.

Case1) The replacing node is itself with 127.0.0.3. 
Since 127.0.0.3 is in normal status, so, existing nodes in the cluster will have 127.0.0.3 in token_metadata

Case2) The replacing node is a node with different IP 127.0.0.4.
127.0.0.4 does not exist in the token_metatda until the replace operation is done.

To make the replacing node see the same tokens in  token_metadata, the replacing node does the same as the rest of the nodes in the cluster.

Reply:
Ok, I get why we have to update tokens early in the same-IP case. We want all nodes to have the same view of the token ring (note that the hash is just an implementation detail here; what is really important is the token ring view being the same).

But, would it be a problem to replace the old tokens early in the different-IP case as well? Would anything bad happen?

Reply:
For the replacing node itself, updating the tokens early is not a problem. The replacing node does not take any read. It also does not have real impact on choosing the source to stream from because neither the replaced node or replacing node will not be considered source. 

I added a comment in the code.